### PR TITLE
Fix IsPrefix and IsSuffix with ignored characters

### LINKED ISSUE #67313: string.StartsWith() displays unexpected behaviour on Unix?
Through an exception in our software I recently got to learn about certain behaviours of string.StartsWith() depending on culture and even operating system. 
Who would have thought that a shorter string can actually start with a longer string 😕 `"A".StartsWith("Aــــــــــ")`.

Now to the question:
Is `"A".StartsWith("A\0")` returning false on Unix expected behaviour?

Opposed to that `"A\0A".EndsWith("AA")`, `"A\0".EndsWith("A")` and `"A".StartsWith("\0A")` all return true on Unix.

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/67313

While performing string `StartsWith` or `EndsWith` operations we try to optimize when having ASCII characters. We apply the optimization only if we don't have characters outside ASCII range nor we don't have characters need special handling (like `\0` which treated as ignored character). We detect the special character by using the internal table `HighCharTable`.  In some cases, we can have the source string be shorter than string we check (e.g. doing something like `"A".StartsWith("A\0")` , we check the remaining character `\0` if it is ASCII but we don't check if it is special character using `HighCharTable`. This cause StartsWith/EndsWith return the wrong result. 
The fix is simply adding the check using `HighCharTable` to ensure the correct result.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs [hunk 1]
```diff
@@ -362,14 +362,14 @@ private unsafe bool StartsWithOrdinalIgnoreCaseHelper(ReadOnlySpan<char> source,
 
                 if (source.Length < prefix.Length)
                 {
-                    if (*b >= 0x80)
+                    if (*b >= 0x80 || HighCharTable[*b])
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs:365]: We cache the char in a local in existing similar places. I assume that it is for good perf. Do we need to do the same here?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs:365]: I don't know but I'll apply the suggestion anyway to follow the same pattern. Thanks!

File: src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.IsPrefix.cs [hunk 2]
```diff
@@ -181,5 +181,12 @@ public void IsPrefix_WithEmptyPrefix_DoesNotValidateOptions()
         {
             IsPrefix(s_invariantCompare, "Hello", "", (CompareOptions)(-1), true, 0);
         }
+
+        [Fact]
+        public void IsPrefixWithAsciiAndIgnoredCharacters()
+        {
+            Assert.StartsWith("A", "A\0");
```
Comment (eerhardt) [inline src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.IsPrefix.cs:188]: Do we want to test ignore case as well?

Comment (tarekgh) [reply src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.IsPrefix.cs:188]: I added the ignore case tests cases. Thanks!

**Review by eerhardt — APPROVED** [review]
Looks good to me.

**Pull request merged**